04-cs-fundamentals/internship/graphs.md

```diff
@@ -0,0 +1,421 @@
+
+![Example graph for an edge list](images/graph-edge-list.png)
+
+Below is an example matrix:
+
+![Example Adjacency matrix](images/graph-adjacency-matrix.png)
+
+In the adjacency matrix above if two nodes matched by the row-column intersection share an edge, that entry in the matri is true.  If they do not share an edge that entry is false.  Notice a couple of things:
+
+- Unless a node has an edge with itself the diagonal from [0][0] to [n-1][n-1] will be false
+- If the graph is undirected the entries across the central diagonal will be mirror images
+  - if (1, 2) is true, then (2, 1) must also be true
+- We would need some way to convert a node in the graph to a number
+
+For a graph G(N, E), where N is the number of nodes and E the number of edges, this solution provides:
+
+- **O(1)** lookup time complexity to find out if any two nodes share an edge
+- **O(N<sup>2</sup>)** space complexity to store this representation
+- **O(1)** time complexity for getting a list of all the neighbors of a node `n` because you need to read the entire row or column.
+
+This method provides a very fast lookup time and not terrible space complexity if the graph is a _dense graph_, in other words if there are a large number of edges compared to nodes.
+
+In the example above we used a matrix of booleans where true indicates an existing edge and false indicates no edge exists connecting the two nodes.  You will often see this represented as a matrix of ones and zeroes where zero indicates no edge exists and 1 indicates it does.  This has the advantage of only requiring one bit per matrix cell.  There are [some strategies](https://github.com/javolio/Bitwise-Adjacency-Matrix) to use this fact to compress the space requirements.  
+
+### Adjacency List
+
+An Adjacency List provides a bit of a middle ground between a list of edges and an adjacency matrix.  In an adjacency list, you have an array with each index representing a node and the content an unordered list of neighboring nodes.  The list can be an array or a linked list.
+
+![Adjacency List Example](images/graph-adjacency-list.png)
+
+This adjacency list gives us:
+
+- **O(d)** lookup time to find out if any two nodes share an edge, where `d` is the _degree_ of a node meaning the number of edges the node has
+  - The maximum value of d would be the case where a node is connected to every other node (assuming no duplicate edges) which is V - 1
+  - This can be a problem in dense graphs
+- O(2E) = **O(E)** space complexity to store this representation because each element of the array stores a list of the edges a node has, twice in the case of an undirected graph.
+- **O(1)** time complexity to find all the neighbors of a particular node because it is an O(1) operation to lookup a node by it's index and the list at that location is a list of all the node's neighbors.
```

Comment:
```suggestion
- **O(1)** time complexity to find list of all of the neighbors of a particular node (because it is an _O(1)_ operation to lookup a node by it's index and the list at that location is a list of all the node's neighbors).
```